pkg/image/layer.go

```diff
@@ -107,44 +99,82 @@ func (l *Layer) Read(catalog *FileCatalog, imgMetadata Metadata, idx int, uncomp
 		types.DockerForeignLayer,
 		types.DockerUncompressedLayer:
 
-		tarFilePath, err := l.uncompressedTarCache(uncompressedLayersCacheDir)
+		err := l.readStandardImageLayer(idx, uncompressedLayersCacheDir, tree)
 		if err != nil {
 			return err
 		}
-
-		l.indexedContent, err = file.NewTarIndex(
-			tarFilePath,
-			layerTarIndexer(tree, l.fileCatalog, &l.Metadata.Size, l, monitor),
-		)
-		if err != nil {
-			return fmt.Errorf("failed to read layer=%q tar : %w", l.Metadata.Digest, err)
-		}
-
 	case SingularitySquashFSLayer:
-		r, err := l.layer.Uncompressed()
-		if err != nil {
-			return fmt.Errorf("failed to read layer=%q: %w", l.Metadata.Digest, err)
-		}
-		// defer r.Close() // TODO: if we close this here, we can't read file contents after we return.
-
-		// Walk the more efficient walk if we're blessed with an io.ReaderAt.
-		if ra, ok := r.(io.ReaderAt); ok {
-			err = file.WalkSquashFS(ra, squashfsVisitor(tree, l.fileCatalog, &l.Metadata.Size, l, monitor))
-		} else {
-			err = file.WalkSquashFSFromReader(r, squashfsVisitor(tree, l.fileCatalog, &l.Metadata.Size, l, monitor))
-		}
+		err := l.readSingularityImage(idx, tree)
 		if err != nil {
-			return fmt.Errorf("failed to walk layer=%q: %w", l.Metadata.Digest, err)
+			return err
 		}
-
 	default:
-		return fmt.Errorf("unknown layer media type: %+v", l.Metadata.MediaType)
+		return fmt.Errorf("unknown layer media type: %+v", mediaType)
 	}
 
 	l.SearchContext = filetree.NewSearchContext(l.Tree, l.fileCatalog.Index)
 
+	return nil
+}
+
+func (l *Layer) readStandardImageLayer(idx int, uncompressedLayersCacheDir string, tree *filetree.FileTree) error {
+	var err error
+	l.Metadata, err = newLayerMetadata(l.layer, idx)
+	monitor := trackReadProgress(l.Metadata)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("layer metadata: index=%+v digest=%+v mediaType=%+v",
+		l.Metadata.Index,
+		l.Metadata.Digest,
+		l.Metadata.MediaType)
+	tarFilePath, err := l.uncompressedTarCache(uncompressedLayersCacheDir)
+	if err != nil {
+		return err
+	}
+
+	l.indexedContent, err = file.NewTarIndex(
+		tarFilePath,
+		layerTarIndexer(tree, l.fileCatalog, &l.Metadata.Size, l, monitor),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to read layer=%q tar : %w", l.Metadata.Digest, err)
+	}
+
 	monitor.SetCompleted()
+	return nil
+}
+
+func (l *Layer) readSingularityImage(idx int, tree *filetree.FileTree) error {
```

Comment:
I see that some of this is a refactoring of what's already there, but we should not be provider-specific code in the main `image` package. We should probably just get rid of the `Read()` call and move appropriate "reading" into the appropriate image providers. Would that help to return more appropriate errors?

Reply:
this does make some sense, but I'm not certain it should be done on this PR (this suggestion needs to be more specific on what the impacts are, which isn't really due for this fix)